mysys/my_largepage.c

```diff
@@ -79,88 +70,361 @@ void my_large_free(uchar* ptr)
+static size_t my_large_page_size= 0;
+
+/* mmap-specific function to determine the size of large pages
+
+This is a fudge as we only use this to ensure that mmap allocations
+are of this size.
+*/
+
+void my_get_large_page_size(void)
+{
+  my_large_page_size= my_getpagesize();
+}
+
+/* mmap(non-Linux) pages allocator  */
+
+uchar* my_large_malloc_int(size_t *size, myf my_flags)
+{
+  uchar* ptr;
+  int mapflag;
+  DBUG_ENTER("my_large_malloc_int");
+
+  mapflag = MAP_PRIVATE | OS_MAP_ANON;
+
+  if (my_use_large_pages && my_large_page_size)
+  {
+    /* Align block size to my_large_page_size */
+    *size= MY_ALIGN(*size, (size_t) my_large_page_size);
+  }
+  /* mmap adjusts the size to the hugetlb page size so no adjustment is needed */
+  ptr = mmap(NULL, *size, PROT_READ | PROT_WRITE, mapflag, -1, 0);
+  if (ptr == (void*) -1)
+  {
+    ptr= NULL;
+    if (my_flags & MY_WME)
+    {
+      fprintf(stderr,
+              "Warning: Failed to allocate %zu bytes from memory."
+              " errno %d\n", *size, errno);
+    }
+  }
+  DBUG_RETURN(ptr);
+}
+#endif /* defined(HAVE_MMAP) && !defined(HAVE_LINUX_LARGE_PAGES) && !defined(_WIN32) */
+
+#ifdef _WIN32
+static size_t my_large_page_size= 0;
+
+/* Windows-specific function to determine the size of large pages */
+
+void my_get_large_page_size(void)
+{
+  DBUG_ENTER("my_get_large_page_size_int");
+
+  my_large_page_size= my_use_large_pages ? GetLargePageMinimum()
+                      : my_getpagesize();
+  DBUG_VOID_RETURN;
+}
+
+/* Windows-specific large pages allocator */
+
+uchar* my_large_malloc_int(size_t *size, myf my_flags)
+{
+  DBUG_ENTER("my_large_malloc_int");
+  void* ptr;
+  DWORD alloc_type= MEM_COMMIT | MEM_RESERVE;
+  size_t orig_size= *size;
+
+  if (my_use_large_pages)
+  {
+    alloc_type|= MEM_LARGE_PAGES;
+    /* Align block size to my_large_page_size */
+    *size= MY_ALIGN(*size, (size_t) my_large_page_size);
+  }
+  ptr= VirtualAlloc(NULL, *size, alloc_type, PAGE_READWRITE);
+  if (!ptr)
+  {
+    if (my_flags & MY_WME)
+    {
+      fprintf(stderr,
+              "Warning: VirtualAlloc(%zu bytes%s) failed; Windows error %lu\n",
+              *size,
+              my_use_large_pages ? ", MEM_LARGE_PAGES" : "",
+              GetLastError());
+    }
+    *size= orig_size;
+    ptr= VirtualAlloc(NULL, *size, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
+    if (!ptr && my_flags & MY_WME)
+    {
+      fprintf(stderr,
+              "Warning: VirtualAlloc(%zu bytes) failed; Windows error %lu\n",
+              *size, GetLastError());
+    }
+  }
 
   DBUG_RETURN(ptr);
 }
 
-/* Linux-specific large pages deallocator */
+/* Windows-specific large pages deallocator */
 
-my_bool my_large_free_int(uchar *ptr)
+my_bool my_large_free_int(void *ptr, size_t size)
 {
   DBUG_ENTER("my_large_free_int");
-  DBUG_RETURN(shmdt(ptr) == 0);
+  /* When RELEASE memory, the size parameter must be 0.
+     Do not use MEM_RELEASE with MEM_DECOMMIT. */
+  if (!VirtualFree(ptr, 0, MEM_RELEASE))
```

Comment:
passing NULL as first parameter to VirtualFree() is handled as an error, which you might want to catch here, because in other APIs this is usually ok.
I have seen Innodb printing both "VirtualAlloc failed with ERROR_INSUFFICIENT_RESOURCES" followed by "VirtualFree failed with ERROR_INVALID_PARAMETER" on allocation errors.  a little ugly.